Fixes onMonthChange called to late on scrolling to the past and implements a PrefetchingWeekViewLoader class

ln-12 opened the pull request:
Same as #86, but using separate branch instead of develop branch to avoid adding more commits. Sorry for that mistake!

entropitor committed:
Move custom font to basicActivity only

entropitor committed:
Merge branch 'pr-3' into develop

jhoobergs committed:
Merge branch 'develop' of github.com:Quivr/Android-Week-View into develop

jhoobergs committed:
Add new repo url

jhoobergs committed:
Code cleanup

jhoobergs committed:
Fix maven upload for annotations

Gradle mvn push script local with small changes to fix javadocs for annotations

jhoobergs committed:
Merge branch 'develop' of github.com:Quivr/Android-Week-View into develop

jhoobergs committed:
Add 2.0.0 release changelog

jhoobergs committed:
Update versionName

jhoobergs committed:
DateTimeInterpreter to interface instead of abstract class

jhoobergs committed:
Update to v2.0.1

jhoobergs committed:
Fix scrolling bug when allday event shown

jhoobergs committed:
v2.0.2

jhoobergs committed:
Update Readme.md

jhoobergs committed:
Merge pull request #16 from tiagomartinho/develop

corrects name of interface in read me instructions

entropitor committed:
Add Unicaen Timetable

jhoobergs committed:
Merge branch 'feature/limit_hour'

Conflicts:
	library/src/main/java/com/alamkanak/weekview/WeekView.java
	library/src/main/res/values/attrs.xml

jhoobergs committed:
Fix min and max time

jhoobergs committed:
update readme

jhoobergs committed:
Some fixes and added timeColumnResolution

pvomhoff committed:
Fix javadoc error in WeekView, set minSdkVersion of sample to 14

pvomhoff committed:
test commit

entropitor committed:
Don't prefetch, leave it up to the WeekViewLoader

jhoobergs committed:
update gradle

jhoobergs committed:
Merge pull request #42 from Quivr/no-prefetch

Don't prefetch, leave it up to the WeekViewLoader

jhoobergs committed:
Merge remote-tracking branch 'origin/develop' into develop

jhoobergs committed:
update version

Commit:
Merge remote-tracking branch 'refs/remotes/Quivr/develop' into develop

jhoobergs committed:
Fix daysBetween

see #49

jhoobergs committed:
Merge pull request #39 from pvomhoff/develop

Fix gradle compile errors

jhoobergs committed:
add some setters and getters

Fixes #43

jhoobergs committed:
Merge remote-tracking branch 'origin/develop' into develop

Commit:
Merge remote-tracking branch 'refs/remotes/origin/develop' into Quivr/develop

Commit:
Merge remote-tracking branch 'refs/remotes/origin/develop' into Quivr/develop

Commit:
fix order of events

sorted events are not in order if there are events longer than one day

eg.

event1: from 21.9.2017 00:00 till 23.9.2017 23:59
event2: from 22.9.2017 10:00 till 22.9.2017 12:00

now the events are sorted and then split/cached

that would result in 4 eventRects

event1-1: from 21.9.2017 00:00 till 21.9.2017 23:59
event1-2: from 22.9.2017 00:00 till 22.9.2017 23:59
event1-3: from 23.9.2017 00:00 till 23.9.2017 23:59
event2: from 22.9.2017 10:00 till 22.9.2017 12:00

That's not in order, event2 happens before event1-3.
So the events should first be split up and cached and then get sorted.

With this change the order would be the following

event1-1: from 21.9.2017 00:00 till 21.9.2017 23:59
event1-2: from 22.9.2017 00:00 till 22.9.2017 23:59
event2: from 22.9.2017 10:00 till 22.9.2017 12:00
event1-3: from 23.9.2017 00:00 till 23.9.2017 23:59

Commit:
add missing cases

extend switch statements to comply the Google Java Style Guide

see
https://google.github.io/styleguide/javaguide.html#s4.8.4.3-switch-default
or http://errorprone.info/bugpattern/MissingCasesInEnumSwitch

jhoobergs committed:
Merge pull request #51 from marunjar/cacheandsort

Cacheandsort

jhoobergs committed:
Merge pull request #52 from marunjar/missingCases

add missing case

jhoobergs committed:
Fix readme link

hermansje committed:
Update README.md

ezmegy committed:
fixed: pastBackgroundColor on current day with minTime set

jhoobergs committed:
Merge pull request #55 from ezmegy/develop

fixed: pastBackgroundColor on current day with minTime set

jhoobergs committed:
increase version

ezmegy committed:
eventPadding minor bug (was always set to default 8) fixed

Commit:
Make EventId on WeekViewEvent generic for custom calendar fetches

jhoobergs committed:
Merge pull request #62 from ezmegy/develop

eventPadding minor bug (was always set to default 8dp) fixed

Commit:
Update Readme.md

jhoobergs committed:
Version 2.2.0

Commit:
fix equality

Commit:
improve string checks

check not only null but also 0-length

append separator ' ' only when location is set, to avoid trailing space
and ellipsizing with small column width

jhoobergs committed:
v2.2.1

jhoobergs committed:
Drag-and-drop implementation (cleanup) (#77)

* Drag-and-drop implementation

* reformatting

* DropListener can be disabled

jhoobergs committed:
Some fixes for new events placement (and other things) (#78)

* Adds support for auto-picking event textcolor based on background color

* Xml support for auto textcolor

* Fixes #29 where adding an event at the end of the day removes all events

* Text color picking interface + example
New event start time adjusted

* Only adjusts the selected position when adding new events

* Only adjusts the selected position when adding new events

* Cloning selected time to prevent changing it in callback

* Adds support for auto-picking event textcolor based on background color

* Xml support for auto textcolor

* Text color picking interface + example
New event start time adjusted

* Fix to prevent losing events when adding new event

* Only adjusts the selected position when adding new events

* Only adjusts the selected position when adding new events

* Cloning selected time to prevent changing it in callback

* Import fix

* Fix for min- and maxTime when adding events

* Fix some issues with NewEvent placement

* remove unused attribute

* reformatting

* update documentation

jhoobergs committed:
update documentation

jhoobergs committed:
v2.2.2

jhoobergs committed:
Add slack link

jhoobergs committed:
Update README.md

jhoobergs committed:
Update README.md

jhoobergs committed:
fix addEventClickListener

jhoobergs committed:
Merge remote-tracking branch 'origin/develop' into develop

jhoobergs committed:
reformatting

jhoobergs committed:
update versionnr

ln-12 committed:
Fixes onMonthChange called to later on scrolling to the past

ln-12 committed:
Implemented a PrefetchingWeekViewLoader class

AndroidDeveloperLB commented:
Does this fix this issue too:
https://github.com/Quivr/Android-Week-View/issues/105
?
Maybe they are the same?
I've created a fork of this repository, and changed it to Kotlin:
https://github.com/AndroidDeveloperLB/Android-Week-View
I hope it won't be too hard to fix it.

ln-12 commented:
Oh, I love Kotlin! ❤️ 

Yep, I think that's exactly the issue I fixed with this PR. The problem is that they just removed the prefetching stuff from the original code without considering all the consequences (e.g. the bug you mentioned). I also created a fork which already includes this fix as the mentainers of this repo don't seem to be very active.
But please recheck it for your situation to see if it works!

AndroidDeveloperLB commented:
Seems it fixes it. But which are the changes that I need to do?
Should I look at this:
https://github.com/Quivr/Android-Week-View/pull/86
Or just here?

AndroidDeveloperLB commented:
Can you please do the needed changes on my fork? Seems quite hard to find the needed changes...

ln-12 commented:
Please don't care about #86, I forgot to do all the relevant commits in a new branch to avoid that new commits are also added there and just did a new PR.

Basically you just need to let Android Studio convert the PrefetchingWeekViewLoader to Kotlin and adjust the bounds stuff. [Here](https://github.com/Quivr/Android-Week-View/pull/89/files) is everything I changed.

AndroidDeveloperLB commented:
OK, I think I got it to work, but why did you call it "PrefetchingWeekViewLoader" ? Why did you wrap it around WeekViewLoader ? And what's the `mPrefetchingPeriod` ?

AndroidDeveloperLB commented:
I think the issue still exists (even in your repository), but in a smaller scale.
See this video:

[device-2018-05-27-134603.zip](https://github.com/Quivr/Android-Week-View/files/2042509/device-2018-05-27-134603.zip)

Not as bad as the original issue, but still... It doesn't show the events of the partially-displayed day till you show more of it.

ln-12 commented:
I implemented it that way as @entropitor suggested it, if I want to contribute that to the lib.

As described in the class file mPrefetchingPeriod is "The amount of periods to be fetched before and after the current period. Must be 1 or greater." That means if you wanna provide data for 1 month, it also requests data for the N months before and after if mPrefetchingPeriod is N.

How did you implement it? Of course you need a prefetching period longer than the amount of visible days as shown [here](https://github.com/mc-12/Android-Week-View/blob/54ca15fee074f0bc934705953445925746b71841/sample/src/main/java/com/alamkanak/weekview/sample/BaseActivity.java#L63).

ln-12 commented:
Wait a second, now I see what the actual problem is. The all day events gap is hidden when there are no more all day events in the visible amount of days. It has nothing to do with this issue.

For that please refer to [this commit](https://github.com/mc-12/Android-Week-View/commit/3d34c84c708d910e0cfdad79f62a32aad3b039ff) and set the property to true.

AndroidDeveloperLB commented:
Hmmmm... I wanted to avoid forcing the developer to set yet another field for this, so I had it set within the class, reducing the hassle.
Probably a bad choice, because you say it should be configurable to the developer.
What happens if I use the default one (with 1 as parameter), always?

About the fix, you can clone my Kotlin fork here :
https://github.com/AndroidDeveloperLB/Android-Week-View

If you know how to improve it , a PR is welcomed :)

Currently there are 2 issues that I still see on this library:
1. No support for RTL in terms of mirroring the UI (meaning order of days will be from right to left, for example
2. Feels a bit slow on old Android devices, compared to Google Calendar app.

AndroidDeveloperLB commented:
About your fix of the all-day, I don't think it's right.
What you wrote will set the all-day row to always appear, even if there aren't any events there.

Also, the issue I've presented is that the event itself isn't shown till you've scrolled enough.

ln-12 commented:
Please try out my suggested fix an set `mWeekView.alwaysShowAllDayEventGap(true);`
I think the event is not shown, because it is hidden inside the collapsed all day header. I just tried it out and it works perfectly well.

AndroidDeveloperLB commented:
How can I clone this commit?
Visiting the repository (here: https://github.com/mc-12/Android-Week-View ) , it says latest commit was 9 days ago

ln-12 commented:
Yeah, that's fine. The [relevant changes](https://github.com/mc-12/Android-Week-View/commit/3d34c84c708d910e0cfdad79f62a32aad3b039ff) were commited back in April.

AndroidDeveloperLB commented:
Well, as I wrote, it always shows the row, even without events:
[device-2018-05-27-151127.zip](https://github.com/Quivr/Android-Week-View/files/2042592/device-2018-05-27-151127.zip)


and if I don't set the flag, the issue I wrote occurs (that events won't be shown till I scroll a bit more).

ln-12 commented:
That's exactly how I think this property is supposed to perform but it definitely does not belong to this PR anymore. The prefetching works fine and has nothing to do with all day events.
I really don't get what the behaviour you want would look like, sorry.

AndroidDeveloperLB commented:
The issue is that for example:

1. I have an all-day even today (2018-5-28), I scroll so that the left day is tomorrow:

![image](https://user-images.githubusercontent.com/5357526/40585946-c36f73f0-61c2-11e8-8664-ac86b7a41180.png)

This is ok. None of those have all day event.

2. If I scroll so that I see entire day of today, I see the event :

![image](https://user-images.githubusercontent.com/5357526/40585953-e0102f4a-61c2-11e8-96e9-b5696ff3aa96.png)

3. However, if I scroll only partially, I don't see it:

![image](https://user-images.githubusercontent.com/5357526/40585957-efbf9250-61c2-11e8-8e95-f35b716c4dff.png)

ln-12 commented:
Ok now i get it, thanks. You shoudl probably take a look at [the lines around here](https://github.com/Quivr/Android-Week-View/blob/2a5f09a6c6ccad3158491acb8c3b05eaf8150b1b/library/src/main/java/com/alamkanak/weekview/WeekView.java#L714) to ix this. But as I said, I think is not related to the prefetching issue and this PR.

AndroidDeveloperLB commented:
Yes that's true. I wrote that it's not the same issue.

AndroidDeveloperLB commented:
Have you looked at my repo?

ln-12 commented:
I don't have time to take a deeper look at it right now. If it works, it's fine I guess. But as I see that you are excessively using the "!!" operator which might be dangerous as it could cause null pointer exceptions. Kotlin provides some syntactical ways to avoid it most of the time.

AndroidDeveloperLB commented:
I converted the previous code. If in the previous code there is no null check, it doesn't need to be here either.
Of course, since the code is huge, it can be shortened and become nicer...

AndroidDeveloperLB commented:
I think I've found an issue in your fork (and the others) :

https://github.com/Quivr/Android-Week-View/issues/106

Can you please check it out?

AndroidDeveloperLB commented:
ok never mind. I've fixed it :)
